operations/research-and-development/product/development-process/mobile-feature-guidelines.md

```diff
@@ -0,0 +1,39 @@
+- For new features, consider a minserverversion check that only executes the code if the server connected to the app is newer than xyz. [See example to disable or not the position field when editing the profile](https://github.com/mattermost/mattermost-mobile/blob/ee4b85edcfee8316db08c31ec5b2a26afb343bd3/app/screens/edit_profile/index.js#L29)
+Bonus: Legacy servers (ie <v5.0) don't have post metadata, don't rely on it.
+
+## #3 Variable Network Conditions
+
+- API requests will fail on mobile because network conditions are much more variable than on desktop. As such defaults need to be carefully chosen to avoid failing requests breaking or blocking core user functionality. Example: if the default for a channel is set to be read only until a permission API request grants the user permissions, this is likely to result in a poor user experience in bad networks. When in doubt, [the default if an API request fails should not change existing behaviour, or a failed API request should either do nothing or notify the user somehow](https://github.com/mattermost/mattermost-mobile/blob/master/app/mm-redux/actions/preferences.ts#L18)
+- [Retries should be added to important API requests](https://github.com/mattermost/mattermost-mobile/blob/master/app/actions/views/channel.js#L607)
+
+## #4 Client Performance Impact
+
+- Mobile devices vary significantly in their computing capabilities, as such performance should be top of mind when developing mobile features.
+- Minimize requests where possible, make requests in parallel if possible, batch dispatches wherever possible. 
+https://github.com/mattermost/mattermost-mobile/blob/master/app/actions/views/user.js#L79
+- Do NOT track requests. Examples:
+  - Here we are fetching the needed data in sequence but holding dispatching any action, and then once we have all the required data we dispatch an action batch. With this we ensure that the redux store is being updated only once and that every mapStateToProps of mounted connected components runs once instead of multiple times, this will then reduce computation time: https://github.com/mattermost/mattermost-mobile/blob/master/app/actions/views/user.js#L51
+  - Here we are fetching data from the server in sequential order and then in parallel, as we need data from one request and then all others are not dependent on each other. The use of Promise.all will help wait on all responses but take into consideration if one of them fails, everything will fail (do add retries if needed). In the end we batch the actions as we do in the example above: https://github.com/mattermost/mattermost-mobile/blob/master/app/actions/views/user.js#L79
+  - Don’t do this: Here we are using a bind function or an action created that is dispatching at least 2 actions in total, one for tracking the start of the request and then if the requests succeeded or failed. That means that calling that function will update the store twice instead of once:
```

Comment:
```suggestion
  - Don’t do this: Here we are using a bind function or an action created that is dispatching at least two actions in total, one for tracking the start of the request and then if the requests succeeded or failed. That means that calling that function will update the store twice instead of once:
```